python/lsst/ip/diffim/dipoleFitTask.py

```diff
@@ -0,0 +1,921 @@
+    DMTN-007.
+
+    Below is a (somewhat incomplete) list of improvements
+    that would be worth investigating, given the time:
+
+    1. evaluate necessity for separate parameters for pos- and neg- images
+    2. only fit background OUTSIDE footprint (DONE) and dipole params INSIDE footprint (NOT DONE)?
+    3. correct normalization of least-squares weights based on variance planes
+    4. account for PSFs that vary across the exposures (should be happening by default?)
+    5. correctly account for NA/masks  (i.e., ignore!)
+    6. better exception handling in the plugin
+    7. better classification of dipoles (e.g. by comparing chi2 fit vs. monopole?)
+    8. (DONE) Initial fast estimate of background gradient(s) params -- perhaps using numpy.lstsq
+    9. (NOT NEEDED - see (1)) Initial fast test whether a background gradient needs to be fit
+    10. (DONE) better initial estimate for flux when there's a strong gradient
+    11. (DONE) requires a new package `lmfit` -- investiate others? (astropy/scipy/iminuit?)
+    """
+
+    # This is just a private version number to sync with the ipython notebooks that I have been
+    # using for algorithm development.
+    _private_version_ = '0.0.4'
+
+    # Create a namedtuple to hold all of the relevant output from the lmfit results
+    resultsOutput = namedtuple('resultsOutput',
+                               ['psfFitPosCentroidX', 'psfFitPosCentroidY',
+                                'psfFitNegCentroidX', 'psfFitNegCentroidY', 'psfFitPosFlux',
+                                'psfFitNegFlux', 'psfFitPosFluxSigma', 'psfFitNegFluxSigma',
+                                'psfFitCentroidX', 'psfFitCentroidY', 'psfFitOrientation',
+                                'psfFitSignaltoNoise', 'psfFitChi2', 'psfFitRedChi2'])
+
+    def __init__(self, diffim, posImage=None, negImage=None):
+        """Algorithm to run dipole measurement on a diaSource
+        @param diffim Exposure on which the diaSources were detected
+        @param posImage "Positive" exposure from which the template was subtracted
+        @param negImage "Negative" exposure which was subtracted from the posImage
+        """
+
+        self.diffim = diffim
+        self.posImage = posImage
+        self.negImage = negImage
+        self.psfSigma = None
+        if diffim is not None:
+            self.psfSigma = diffim.getPsf().computeShape().getDeterminantRadius()
+        self.log = pexLog.Log(pexLog.Log.getDefaultLog(),
+                              'lsst.ip.diffim.DipoleFitAlgorithm', pexLog.Log.INFO)
+
+        import lsstDebug
+        self.debug = lsstDebug.Info(__name__).debug
+
+    def genBgGradientModel(self, in_x, pars=()):
+        """Generate gradient model (2-d array) with up to 2nd-order polynomial
+        @param in_x Provides the input x,y grid upon which to compute the gradient
+        @param pars Up to 6 floats for up to 6 2nd-order 2-d polynomial gradient parameters.
+        If emtpy, do nothing (for speed).
```

Comment:
First, I'll make the default `pars=None`. The first case (`pars=()`) should never occur in my code, but `pars=(None)` might occur. Should I make the `pars=(None)` the default? Seems like a strange option to me.

To answer your latter question - I'd rather have the logic of checking the contents tuple be in one place rather than prior to each time that I call this function.